stonesoup/dataassociator/neighbour.py

```diff
@@ -106,3 +111,92 @@ def associate(self, tracks, detections, time):
         associations = max(joint_hypotheses)
 
         return associations
+
+
+class GNNWith2DAssignment(DataAssociator):
+    """Global Nearest Neighbour Associator
+
+    Associates detections to a predicted state using the
+    Global Nearest Neighbour method, utilising a 2D matrix of
+    distances and a "shortest path" assignment algorithm.
+    """
+
+    hypothesiser = Property(
+        Hypothesiser,
+        doc="Generate a set of hypotheses for each prediction-detection pair")
+
+    def associate(self, tracks, detections, time):
+        """Associate a set of detections with predicted states.
+
+        Parameters
+        ----------
+        tracks : set of :class:`Track`
+            Current tracked objects
+        detections : set of :class:`Detection`
+            Retrieved measurements
+        time : datetime
+            Detection time to predict to
+
+        Returns
+        -------
+        dict
+            Key value pair of tracks with associated detection
+        """
+
+        # Convert sets to indexable lists
+        tracks = list(tracks)
+        detections = list(detections)
+
+        # Generate 2d array "matrix" of hypotheses mapping track to detection
+        hypothesis_matrix = np.empty(
+            [len(tracks), len(detections) + len(tracks)],
+            SingleDistanceHypothesis)
+        for i in range(len(tracks)):
+            row = np.empty(
+                [len(detections) + len(tracks)], SingleDistanceHypothesis)
+            track_hypotheses = self.hypothesiser.hypothesise(
+                tracks[i], detections, time)
+            for hypothesis in track_hypotheses:
+                if isinstance(hypothesis.measurement, MissedDetection):
+                    row[len(detections) + i] = hypothesis
+                else:
+                    row[int(detections.index(
+                        hypothesis.measurement))] = hypothesis
+            hypothesis_matrix[i] = row
+
+        probability_flag = isinstance(self.hypothesiser, PDAHypothesiser)
```

Comment:
This should check the types of _hypothesis_ as this will work more generically with other _hypothesisers_. Maybe something like (not tested):
```suggestion
        if all(isinstance(hypothesis, ProbabilityHypothesis) for hypothesis in hypothesis_matrix):
            probability_flag = True
        elif all(isinstance(hypothesis, DistanceHypothesis) for hypothesis in hypothesis_matrix):
            probability_flag = False
        else:
            raise RuntimeError("2d assignment doesn't support mixed types of Hypothesis")
```